smithy-model/src/main/java/software/amazon/smithy/model/validation/AbstractValidator.java

```diff
@@ -28,48 +28,104 @@ public String getName() {
     }
 
-    protected final ValidationEvent warning(Shape shape, FromSourceLocation location, String message) {
-        return createEvent(Severity.WARNING, shape, location, message);
+    protected final ValidationEvent warning(
+            Shape shape,
+            FromSourceLocation location,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(Severity.WARNING, shape, location, message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent note(Shape shape, String message) {
-        return createEvent(Severity.NOTE, shape, shape.getSourceLocation(), message);
+    protected final ValidationEvent note(
+            Shape shape,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(Severity.NOTE, shape, shape.getSourceLocation(), message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent note(Shape shape, FromSourceLocation location, String message) {
-        return createEvent(Severity.NOTE, shape, location, message);
+    protected final ValidationEvent note(
+            Shape shape,
+            FromSourceLocation location,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(Severity.NOTE, shape, location, message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent createEvent(Severity severity, Shape shape, String message) {
-        return createEvent(severity, shape, shape.getSourceLocation(), message);
+    protected final ValidationEvent createEvent(
+            Severity severity,
+            Shape shape,
+            String message,
+            String... additionalEventIdParts
+    ) {
+        return createEvent(severity, shape, shape.getSourceLocation(), message, additionalEventIdParts);
     }
 
-    protected final ValidationEvent createEvent(Severity severity, Shape shape, FromSourceLocation loc, String msg) {
-        return createEvent(ValidationEvent.builder().severity(severity).message(msg)
-                                   .shapeId(shape.getId()).sourceLocation(loc.getSourceLocation()));
+    protected final ValidationEvent createEvent(
+            Severity severity,
+            Shape shape,
+            FromSourceLocation loc,
+            String msg,
+            String... additionalEventIdParts
+    ) {
+        return ValidationEvent.builder()
+                .severity(severity)
+                .message(msg)
+                .shapeId(shape.getId())
+                .sourceLocation(loc.getSourceLocation())
+                .id(assembleFullEventId(additionalEventIdParts))
+                .build();
     }
 
-    private ValidationEvent createEvent(ValidationEvent.Builder builder) {
-        return builder.id(getName()).build();
+    protected final String assembleFullEventId(String... additionalEventIdParts) {
```

Comment:
We should optimize this method since it'll be used a ton. First check if the length is 1, and if so, just return the first entry. If > 1, then use a StringBuilder and build up the result.

Reply:
optimized in latest commit